---
title: "R Notebook"
output: html_notebook
---

### Quiz 2

```{r load}
library(SummarizedExperiment)
library(Biobase)
library(GenomicRanges)
library(broom)
```

### PCA - Question 1

How to calculate the percent of variance explained?

* https://stats.stackexchange.com/questions/22569/pca-and-proportion-of-variance-explained
* http://jtleek.com/genstats/inst/doc/02_03_dimension-reduction.html#look-at-the-percent-variance-explained
* http://genomicsclass.github.io/book/pages/svd.html

$$
\frac{d_{ii}}{\sum_{j} d^2_{jj}}
$$

```{r}
con =url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/montpick_eset.RData")
load(file=con)
close(con)
mp = montpick.eset
pdata=pData(mp)
edata=as.data.frame(exprs(mp))
fdata = fData(mp)
```

```{r}

edata2 = log2(edata + 1)
edata3 = edata2 - rowMeans(edata2)  
svd1 = svd(edata) #need to repeat for each of edata1,2,3
names(svd1)
```

Just for fun, plot the relative size of singular values.

```{r}
plot(svd1$d, ylab="Singular Value", col=2)
```

```{r}
plot(svd1$d^2/sum(svd1$d^2),ylab="Percent Variance Explained",col=2)
```

```{r}
svd2 = svd(edata2) 
svd3 = svd(edata3)
```



```{r}
plot(svd2$d^2/sum(svd2$d^2),ylab="Percent Variance Explained- edata2",col=2)

```

```{r}
plot(svd3$d^2/sum(svd3$d^2),ylab="Percent Variance Explained- edata3",col=2)

```

One thing to remember, SVD can flip signs.  

#### K-means clustering + SVD correlation - Question 2 

```{r}
#log2 transform + rowmeans
edataQ2 = log2(edata + 1) 
edataQ2 = edataQ2 - rowMeans(edataQ2)
```

```{r clusterQ2}
set.seed(333)
kmQ2 = kmeans(t(edataQ2), 2)
svdQ2 = svd(edataQ2)
```

```{r}
cor(svdQ2$v[,1], kmQ2$cluster)
```

### Question 3 Linear models

Fit a linear model related the first gene's counts to the number of technical replicates.  Treat the number of technical replicates as a factor.

See 

* http://jtleek.com/genstats/inst/doc/02_13_batch-effects.html
* http://jtleek.com/genstats/inst/doc/02_08_poverty-example.html
* http://jtleek.com/genstats/inst/doc/02_11_many-regressions.html
* http://jtleek.com/genstats/inst/doc/02_06_galton-example.html
* https://genomicsclass.github.io/book/pages/adjusting_with_linear_models.html
* https://genomicsclass.github.io/book/pages/eda_with_pca.html
* https://genomicsclass.github.io/book/pages/expressing_design_formula.html


```{r}
con =url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/bodymap_eset.RData")
load(file=con)
close(con)
bm = bodymap.eset
edata = exprs(bm)
pdata_bm=pData(bm)
```

```{r}
#gene1 = as.matrix(edata[1,])
#reps = as.matrix(as.factor(pdata$num.tech.reps))
#lm <- lm.fit(gene1 ~ reps)

mod = model.matrix(~ as.factor(num.tech.reps), data=pdata_bm)
fit = lm.fit(mod, t(edata))
```

Plot the data for this gene versus the covariate. Can you think of why this model might not fit well?

* The data are right skewed.

Right skewed distributions would have a long tail on the RHS. They would also show Mode <= Median <= Mean.  There is so little data here I have mixed feelings as to whether I'd categorize this as a true right skewed distribution.  Also, we're just talking about a histogram of counts for `num.tech.reps` so I'm not sure if that qualifies for "the data".  

* There is only one data point with a value of 6 so it is likely that the estimated value for that number of technical replicates is highly variable.

Certainly the value for 6 rechnical replicates would be more variable. I'm not sure that explains why the fit is off.  So far, I'd pick this answer. 

* There may be different numbers of counts for different numbers of technical replicates.

What does "numbers of counts" mean?  I would assume absolute count values are divided by the numbder of technical replicates. I'm not sure what this answer is stating. 

* The difference between 2 and 5 technical replicates is not the same as the difference between 5 and 6 technical replicates.

There is certainly a gap in data between 2 and 5. 5 and 6 are much closer  Does that explain why the fit might be off? No. 

```{r}
#3 rows by 52k cols. Look in the first column for data about the first gene
dim(fit$coefficients)

#why three coefficients?
fit$coefficients[,1]

#what should I plot? Historgram or linear model? In video lecture about correcting for batch effects a histogram is plotted first. 

#Update: it's a histogram of LM coefficients (for the outcome variable cancer?).  It's just their frequency. How many coefficients for different genes take on different values (i.e. slopes).  Most values have a zero slope (the expression count is uncorrelated with the outcome).  Some are highly negatively or positively correlated. 

#You don't really fit data if you using a histogram.  Assume they want a linear plot. 

# histogram would likely be gene expression level on y axis and num technical replicates n x-axis.  This would be the same setup for a linear model. 
plot(pdata_bm$num.tech.reps, edata[1,])
```
Just for fun,  let's plot this with ggplot

```{r}
library(ggplot2)
```

```{r}
df <- data.frame(ensg = edata[1,], num.tech.reps = pdata_bm$num.tech.reps)
ggplot(data=df, aes(y=ensg, x = num.tech.reps)) + geom_point() + geom_smooth(method="lm", color="red", linetype=2, formula="y~x")
```

4. Fit a linear model relating the first geneâ€™s counts to the age of the person and the sex of the samples. What is the value and interpretation of the coefficient for age?

Let's try fitting all genes just for fun.  
```{r}
options(na.action='na.pass')
m4  <- model.matrix(~age + as.factor(gender), data=pdata_bm)
#ft4 <- lm.fit(m4, t(edata), na.action='na.pass')
lm4a <- lm(t(edata) ~ age + gender, data=pdata_bm, na.action = na.omit)
```

Now can we fit just the first gene? 

```{r}
gene1_4 <- edata[1,] 
lm4 <- lm(gene1_4 ~ age + gender, data=pdata_bm, na.action = na.omit)
tidy(lm4)
```


