---
title: "R Notebook"
output: html_notebook
---

# Quiz 1

## Setup

```{r install_packages, results="hide", warning=FALSE, eval=FALSE}
install.packages(c("devtools"))
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")


BiocManager::install("Biobase")
BiocManager::install("GenomicRanges")
BiocManager::install("SummarizedExperiment")
```

```{r load}
library(SummarizedExperiment)
library(Biobase)
library(GenomicRanges)
```


#### Testing the cache

```{r setup, eval=TRUE}
knitr::opts_chunk$set(cache=FALSE)
```

```{r}
x = rnorm(10)
plot(x,pch=19,col="dodgerblue")
```

```{r}
y = rbinom(20,size=1,prob=0.5)
table(y)
```

#### Working with SummarizedExperiment obj

```{r}
library(Biobase)
library(GenomicRanges)
data(sample.ExpressionSet, package = "Biobase")
se = makeSummarizedExperimentFromExpressionSet(sample.ExpressionSet)
```

```{r}
# genes (rows) x samples (columns).  Cells are expression values (normalized?)
assay(se)
```

```{r}
# phenotype data
colData(se)
```

```{r}
# shows the name of the genes that are being expressed. No data.  Returns a data fram
rowData(se)
```
```{r}
rowRanges(se)
```

```{r se_helpers, results="hide"}
dim(se)
dimnames(se)
```

### Experimental Design - Bottomly and Bodymap

```{r load_bot_bm}
con =url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/bottomly_eset.RData")
load(file=con)
close(con)
bot = bottomly.eset
pdata_bot=pData(bot)

con =url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/bodymap_eset.RData")
load(file=con)
close(con)
bm = bodymap.eset
pdata_bm=pData(bm)
```

### Graphing Bodymap

```{r}
library(plotrix)
pie3D(pdata_bm$num.tech.reps,labels=pdata_bm$tissue.type)
```

### Bodymap Heatmaps

Which of the following code chunks will make a heatmap of the 500 most highly expressed genes (as defined by total count), without re-ordering due to clustering?

```{r}
edata = exprs(bm)
row_sums = rowSums(edata)
edata = edata[order(-row_sums),] #this reorders the rows
index = 1:500
heatmap(edata[index,],Rowv=NA,Colv=NA)
```

```{r}
edata = exprs(bm)
row_sums = rowSums(edata)
index = which(rank(row_sums) < 500 ) #no row_sums are less than 500. 
#dim(index)
heatmap(edata[index,],Colv=NA)
```

```{r}
edata = exprs(bm)
row_sums = rowSums(edata)
edata = edata[order(row_sums),] #reorder rows. Ascending order. 
index = which(rank(-row_sums) < 500)
heatmap(edata[index,],Rowv=NA, Colv=NA) #not sure why this failed. 
```

```{r}
edata = exprs(bm)
row_sums = rowSums(edata)
index = which(rank(-row_sums) < 500 ) #rank has nothing to do with "rank" in linear algebra. It give you the indexes of the matrix in ascending order. 
heatmap(edata[index,],Rowv=NA,Colv=NA)
```
### Bodymap graphs

Make an MA-plot of the first sample versus the second sample using the **log2** transform (hint: you may have to add 1 first) and the **rlog** transform from the DESeq2 package. How are the two MA-plots different? Which kind of genes appear most different in each plot?


```{r}
con =url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/bodymap_eset.RData")
load(file=con)
close(con)
bm = bodymap.eset
pdata = pData(bm)
edata = exprs(bm)
fdata = fData(bm)

#Now make the graph. 
#sample1 vs. #sample2 with log2 transform. 
par(pch=19)
```

#### MA-plots

Make an MA-plot of the first sample versus the second sample using the `log2` transform (hint: you may have to add mple1 + 11 first) and the `rlog` transform from the DESeq2 package. How are the two MA-plots different? Which kind of genes appear most different in each plot?

https://en.wikipedia.org/wiki/MA_plot

```{r}
edata = exprs(bm)

sample1 = edata[,1]
sample2 = edata[,2]

sdiff = log2(sample1 + 1) - log2(sample2 + 1)
ssum  = log2(sample1 + 1) + log2(sample2 + 1)

plot(ssum, sdiff, col=2, xlab="A", ylab="M")
title("MA Plot")

```
### DESeq2 - rlog

* https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1.1_RNASeqLab.html


```{r install_deseq2, eval=FALSE}
BiocManager::install("DESeq2")
```

```{r}
library(DESeq2)
```
```{r}
# https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1.1_RNASeqLab.html
#This produces -Inf for zero values but might work.
#de_edata = rlog(data.matrix(edata))

#we could just add one to all cells. 
de_edata = rlog(data.matrix(edata+1))

sde1 = de_edata[,1]
sde2 = de_edata[,2]

ssum  = sde1 + sde2
sdiff = sde1 - sde2

plot(ssum, sdiff, col=2, xlab="A", ylab="M")
title("MA Plot")

```


Todo:

* convert edata to a matrix?
  + data.matrix
  + DESeqDataSetFromMatrix
  + consider watching data exploration videos again. 
* make another MA plot



Did both of the first two samples have the same phenotye / labels? Nope. 

```{r}
colnames(pdata)
ps1 = pdata[1,]
ps2 = pdata[2,]
ps1$tissue.type
ps2$tissue.type
```

### Cluster the Data

```{r}
con =url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/montpick_eset.RData")
load(file=con)
close(con)
mp = montpick.eset
pdata=pData(mp)
edata=as.data.frame(exprs(mp))
fdata = fData(mp)
```

```{r eval=FALSE}
install.packages('dendextend')
install.packages("rafalib")
```


```{r load_de}
  library(dendextend)
  library(rafalib)
```

```{r}
dim(pdata) # sample x pheno type
dim(edata)  # gene x sample
dim(fdata) #gene
colnames((pdata)) #pdata has the study.
unique(pdata$study) #only two different values. 
pdata$study[0:10]
tail(pdata$study, 10) 
```
```{r}
is.data.frame(fdata)
fdata$gene[1] #confuses me.
```

#### Dendograms

```{r}
edata2 = edata[rowMeans(edata) >= 100,]
edata3 = log2(edata+1)
```

```{r}
dist1   = dist(t(edata))
hclust1 = hclust(dist1)

```

```{r}
ccolors = with(pdata, ifelse(study=='Montgomery', 1, 2))
myplclust(hclust1, labels = pdata$study, lab.col = ccolors, hang = 0.1, sub="Clustering - no Normalization") 
```

```{r}
dist2   = dist(t(edata2))
hclust2 = hclust(dist2)

```

```{r}
ccolors = with(pdata, ifelse(study=='Montgomery', 1, 2))
myplclust(hclust2, labels = pdata$study, lab.col = ccolors, hang = 0.1, sub="Clustering - filter out rows with less raw counts") 

```

```{r}
dist3   = dist(t(edata3))
hclust3 = hclust(dist3)

```
```{r}
ccolors = with(pdata, ifelse(study=='Montgomery', 1, 2))
myplclust(hclust3, labels = pdata$study, lab.col = ccolors, hang = 0.1, sub="Clustering - rows transformed by log2 + 1") 
```
```{r}
#doesn't appear to d well. 
cutree(hclust3, 2)

```


#### K-means Clustering

```{r}
con =url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/montpick_eset.RData")
load(file=con)
close(con)
mp = montpick.eset
pdata=pData(mp)
edata=as.data.frame(exprs(mp))
fdata = fData(mp)
```


```{r}
set.seed(1235)
edl = log2(edata+1)
km1 = kmeans(t(edl), centers=2) #I think I need columns to be clustered (not rows)
matplot(km1$centers, col=1:2, type="l", lwd=2) #I guess you graph every dim of each km. 
```

```{r}
km1$cluster
table(km1$cluster)
```

```{r}
ground_truth = with(pdata, ifelse(study=='Montgomery', 1, 2))

#this seems to work but I need to count the FALSE values. 
ground_truth == km1$cluster



```


```{r}
colramp = colorRampPalette(c(3,"white",2))(9)
#this just shows the cluster (not the study)
heatmap(as.matrix(edl)[order(km1$cluster),], col=colramp, Colv=NA, Rowv=NA)
```


### Bonus exploratory data analysis

```{r}

table(pdata$gender, pdata$tissue.type) #super sparse. 
table(pdata$gender, pdata$race)

summary(edata)

```

Three tables of genomic data science

* expression data
* data about patient / sample phenotypes
* data about each gene (pathways, snps?, etc.)

In R the common classes that allow you to work with the three tables

* ExpressionSet
  - assayData
  - phenoData
  - featureData


```{r}
#assume you retrieved bm and bot datasets
# these classes use ExpressionSet and AnnotatedDataFrame
f_bm = featureData(bm)[0:10,]
varMetadata(f_bm)
featureNames(f_bm)

```

* SummarizedExperiment
  - assays
  - colData
  - rowData
  

```{r}
#assume you loaded se
# SummarizedExperiment
fd_se = rowData(se)[0:10,]
dim(fd_se) #500 x 0 - no data
rownames(fd_se)

```





## Session information

Here is the session information 

```{r session_info}
devtools::session_info()
```

It is also useful to compile the time the document was processed. This document was processed on: `r Sys.Date()`.

## R Notebook Commands

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

